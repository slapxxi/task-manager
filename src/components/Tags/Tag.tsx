import { ID, Tag as ITag } from '@local/types';
import React from 'react';
import styles from './styles.css';

interface Props {
  tag: ITag;
  parentId?: string;
  selected?: boolean;
  onSelect?: (isSelected: boolean, tag: ITag) => void;
}

function Tag({ tag, parentId, selected, onSelect }: Props) {
  const uniqueId = generateId(parentId || '', tag.id);

  function handleSelect(e: React.ChangeEvent<HTMLInputElement>) {
    if (onSelect) {
      onSelect(e.currentTarget.checked ? true : false, tag);
    }
  }

  return (
    <>
      <input
        type="checkbox"
        className={styles.checkbox}
        id={uniqueId}
        onChange={handleSelect}
        checked={selected}
        data-testid="checkbox"
      />
      <label htmlFor={uniqueId} className={styles.label}>
        {tag.name}
      </label>
    </>
  );
}

function generateId(parentId: ID, id: ID) {
  return `${parentId}-${id}`;
}

export default React.memo(Tag);
