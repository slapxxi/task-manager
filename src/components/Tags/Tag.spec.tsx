import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import Tag from './Tag';

const tag = {
  id: 'test',
  name: 'test',
};

it('renders', () => {
  const { container } = render(<Tag tag={tag} selected={false} />);
  expect(container).toMatchSnapshot();
});

it('clicking does not change the attribute', () => {
  const { container, getByText } = render(<Tag tag={tag} selected={false} />);
  fireEvent.click(getByText('test'));
  expect(container).toMatchSnapshot();
});

it('invokes callback when label clicked', () => {
  const spy = jest.fn();
  const { getByText } = render(
    <Tag tag={tag} onSelect={spy} selected={false} />,
  );
  fireEvent.click(getByText('test'));
  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith(true, tag);
});

it('invokes callback when checkbox clicked', () => {
  const spy = jest.fn();
  const { getByTestId } = render(
    <Tag tag={tag} onSelect={spy} selected={false} />,
  );
  fireEvent.click(getByTestId('checkbox'));
  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith(true, tag);
});

it('invokes callback with correct value', () => {
  const spy = jest.fn();
  const { getByTestId } = render(<Tag tag={tag} onSelect={spy} selected />);
  fireEvent.click(getByTestId('checkbox'));
  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith(false, tag);
});
