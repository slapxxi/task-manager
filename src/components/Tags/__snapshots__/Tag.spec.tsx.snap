// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clicking does not change the attribute 1`] = `
<div>
  <input
    class="checkbox"
    data-testid="checkbox"
    id="-test"
    type="checkbox"
  />
  <label
    class="label"
    for="-test"
  >
    test
  </label>
</div>
`;

exports[`renders 1`] = `
<div>
  <input
    class="checkbox"
    data-testid="checkbox"
    id="-test"
    type="checkbox"
  />
  <label
    class="label"
    for="-test"
  >
    test
  </label>
</div>
`;
